File: utils/apiUrl.js
const path = require('path');
const dotenv = require('dotenv');
const axios = require("axios");

// 환경 변수 로딩
const envPath = path.resolve(__dirname, '../.env');
dotenv.config({ path: envPath });

const API_KEY = process.env.WEATHER_API_KEY;
const BASE_URL = "http://api.openweathermap.org/data/2.5/forecast";

// 날씨 데이터를 가져오는 함수
async function fetchWeatherData(city) {
  try {
    const encodedCity = encodeURIComponent(city);
    const url = `${BASE_URL}?q=${encodedCity}&appid=${API_KEY}&units=metric&lang=kr`;

    const response = await axios.get(url);
    
    // 현재 날씨 데이터 수정
    const currentWeather = {
      도시: response.data.city.name,
      현재기온: response.data.list[0].main.temp,
      최고기온: response.data.list[0].main.temp_max,
      최저기온: response.data.list[0].main.temp_min,
      습도: response.data.list[0].main.humidity,
      날씨설명: response.data.list[0].weather[0].description,
      강수량: response.data.list[0].rain ? response.data.list[0].rain['3h'] : 0
    };

    // 내일 날씨 데이터 처리
    const tomorrowWeather = {
      최고기온: Math.max(...response.data.list.slice(8, 16).map(item => item.main.temp_max)),
      최저기온: Math.min(...response.data.list.slice(8, 16).map(item => item.main.temp_min)),
      습도: response.data.list[8].main.humidity,
      날씨설명: response.data.list[8].weather[0].description,
      강수량: response.data.list[8].rain ? response.data.list[8].rain['3h'] : 0
    };

    // 주간 날씨 데이터 처리
    const weeklyWeather = [];
    for (let i = 0; i < 40; i += 8) {
      const dayData = response.data.list.slice(i, i + 8);
      const date = new Date(dayData[0].dt * 1000);
      
      weeklyWeather.push({
        날짜: date.toLocaleDateString('ko-KR'),
        최고기온: Math.max(...dayData.map(item => item.main.temp_max)),
        최저기온: Math.min(...dayData.map(item => item.main.temp_min)),
        습도: dayData[0].main.humidity,
        날씨설명: dayData[0].weather[0].description,
        강수량: dayData[0].rain ? dayData[0].rain['3h'] : 0
      });
    }

    return {
      current: currentWeather,
      tomorrow: tomorrowWeather,
      weekly: weeklyWeather,
      raw: response.data
    };
  } catch (error) {
    console.error("날씨 데이터를 가져오는데 실패했습니다:", error);
    throw error;
  }
}

module.exports = {
  API_KEY,
  BASE_URL,
  fetchWeatherData
};

File: utils/apiUrl.py
import os
from dotenv import load_dotenv
import aiohttp
import json
import urllib.parse

# 환경 변수 로딩
load_dotenv()

API_KEY = os.getenv('WEATHER_API_KEY')
BASE_URL = "http://api.openweathermap.org/data/2.5/forecast"

# 한국 주요 도시 매핑 (한글:영문)
KOREAN_CITIES = {
    "서울": "Seoul",
    "부산": "Busan",
    "인천": "Incheon",
    "대구": "Daegu",
    "대전": "Daejeon",
    "광주": "Gwangju",
    "울산": "Ulsan",
    "제주": "Jeju"
}

async def fetchWeatherData(city):
    try:
        # 한글 도시명을 영문으로 변환
        english_city = KOREAN_CITIES.get(city, city)
        
        async with aiohttp.ClientSession() as session:
            params = {
                'q': f"{english_city},KR",  # 국가 코드 추가
                'appid': API_KEY,
                'units': 'metric',
                'lang': 'kr'
            }
            
            async with session.get(BASE_URL, params=params) as response:
                if response.status != 200:
                    error_data = await response.json()
                    raise ValueError(f"날씨 API 오류: {error_data.get('message', '알 수 없는 오류')}")
                    
                raw_data = await response.json()
                processed_data = processWeatherData(raw_data)
                
                # 한글 도시명 추가
                processed_data['korean_name'] = city if city in KOREAN_CITIES else english_city
                return processed_data
                
    except Exception as e:
        print(f"날씨 데이터를 가져오는데 실패했습니다: {str(e)}")
        raise e

def processWeatherData(weatherData):
    try:
        list_data = weatherData.get('list', [])
        if not list_data:
            raise ValueError("날씨 데이터가 비어있습니다")

        # 현재, 내일, 주간 데이터 처리
        current_data = list_data[0]
        tomorrow_data = list_data[1]
        weekly_data = list_data[1:6]  # 5일치 데이터

        # 데이터 형식 변환
        processed = {
            'current': {
                'avg temp': current_data['main']['temp'],
                'max temp': current_data['main']['temp_max'],
                'min temp': current_data['main']['temp_min'],
                'rainFall': current_data.get('rain', {}).get('3h', 0)
            },
            'tomorrow': {
                'avg temp': tomorrow_data['main']['temp'],
                'max temp': tomorrow_data['main']['temp_max'],
                'min temp': tomorrow_data['main']['temp_min'],
                'rainFall': tomorrow_data.get('rain', {}).get('3h', 0)
            },
            'weekly': [
                {
                    'avg temp': day['main']['temp'],
                    'max temp': day['main']['temp_max'],
                    'min temp': day['main']['temp_min'],
                    'rainFall': day.get('rain', {}).get('3h', 0)
                }
                for day in weekly_data
            ],
            'raw': weatherData  # 원본 데이터도 유지
        }
        
        return processed
        
    except Exception as e:
        print(f'날씨 데이터 처리 중 오류 발생: {str(e)}')
        raise e 